boa/src/value/rcstring.rs

```diff
@@ -1,114 +1,448 @@
 use crate::gc::{empty_trace, Finalize, Trace};
-
 use std::{
+    alloc::{alloc, dealloc, Layout},
     borrow::Borrow,
-    fmt::{self, Display},
+    cell::Cell,
+    hash::{Hash, Hasher},
+    marker::PhantomData,
     ops::Deref,
-    rc::Rc,
+    ptr::{copy_nonoverlapping, NonNull},
 };
 
-#[derive(Debug, Finalize, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct RcString(Rc<str>);
+/// The inner representation of a [`JsString`].
+#[repr(C)]
+struct Inner {
+    /// The utf8 length, the number of bytes.
+    len: usize,
 
-unsafe impl Trace for RcString {
-    empty_trace!();
+    /// The number of references to the string.
+    ///
+    /// When this reaches `0` the string is deallocated.
+    refcount: Cell<usize>,
+
+    /// An empty array which is used to get the offset of string data.
+    data: [u8; 0],
```

Comment:
Could we use something like [`NonNull`](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html) for this?

Reply:
But we aren't pointing to any other allocation, it's only one allocation for len, refcount and string data, and size of `NonNull` is size of pointer while size of `[u8; 0]` is zero, we just use it to get the offset of the string bytes. by calling `data.as_ptr()`. This pattern is also used in C/C++ [Flexible array member](https://en.wikipedia.org/wiki/Flexible_array_member)